runtime/stdlib/account.go

```diff
@@ -2297,20 +2297,156 @@ func newAuthAccountStorageCapabilitiesGetControllerFunction(
 				panic(errors.NewUnreachableError())
 			}
 
-			capabilityController := getCapabilityController(inter, address, capabilityIDValue)
-			storageCapabilityController, ok := capabilityController.(*interpreter.StorageCapabilityControllerValue)
-			if !ok {
+			capabilityID := uint64(capabilityIDValue)
+
+			referenceValue := getStorageCapabilityControllerReference(inter, address, capabilityID)
+			if referenceValue == nil {
 				return interpreter.Nil
 			}
 
-			referenceValue := interpreter.NewEphemeralReferenceValue(
+			return interpreter.NewSomeValueNonCopying(inter, referenceValue)
+		},
+	)
+}
+
+var storageCapabilityControllerReferencesArrayStaticType = interpreter.VariableSizedStaticType{
+	Type: interpreter.ReferenceStaticType{
+		BorrowedType: interpreter.PrimitiveStaticTypeStorageCapabilityController,
+	},
+}
+
+func newAuthAccountStorageCapabilitiesGetControllersFunction(
+	gauge common.MemoryGauge,
+	addressValue interpreter.AddressValue,
+) interpreter.FunctionValue {
+	address := addressValue.ToAddress()
+	return interpreter.NewHostFunctionValue(
+		gauge,
+		sema.AuthAccountStorageCapabilitiesTypeGetControllersFunctionType,
+		func(invocation interpreter.Invocation) interpreter.Value {
+
+			inter := invocation.Interpreter
+
+			// Get path argument
+
+			targetPathValue, ok := invocation.Arguments[0].(interpreter.PathValue)
+			if !ok || targetPathValue.Domain != common.PathDomainStorage {
+				panic(errors.NewUnreachableError())
+			}
+
+			// Get capability controllers iterator
+
+			nextCapabilityID, count :=
+				getPathCapabilityControllerIDsIterator(inter, address, targetPathValue)
+
+			var capabilityControllerIndex uint64 = 0
+
+			return interpreter.NewArrayValueWithIterator(
 				inter,
-				false,
-				storageCapabilityController,
-				sema.StorageCapabilityControllerType,
+				storageCapabilityControllerReferencesArrayStaticType,
+				common.Address{},
+				count,
+				func() interpreter.Value {
+					if capabilityControllerIndex >= count {
+						return nil
+					}
+					capabilityControllerIndex++
+
+					capabilityID, ok := nextCapabilityID()
+					if !ok {
+						return nil
+					}
+
+					referenceValue := getStorageCapabilityControllerReference(inter, address, capabilityID)
+					if referenceValue == nil {
+						panic(errors.NewUnreachableError())
+					}
+
+					return referenceValue
+				},
 			)
+		},
+	)
+}
 
-			return interpreter.NewSomeValueNonCopying(inter, referenceValue)
+// the AccountKey in ` forEachController(forPath: StoragePath, _ function: ((&StorageCapabilityController): Bool))`
+var authAccountStorageCapabilitiesForEachControllerCallbackTypeParams = []sema.Type{
+	&sema.ReferenceType{
+		Type: sema.StorageCapabilityControllerType,
+	},
+}
+
+func newAuthAccountStorageCapabilitiesForEachControllerFunction(
```

Comment:
What is the behavior if a controller is added or removed during iteration? IMO the same model we use for storage iteration (and will use for attachment iteration) would be fine, i.e. to panic if mutation occurs. Either way we should have some tests for it.

Reply:
That's a good question/point and was also brought up by Janez in the PR review session yesterday.

I'm currently working on improving the behaviour to allow mutation while iterating, i.e. issuing new controllers and deleting existing controllers, and adding tests. As discussed yesterday, it probably is too limiting to reject mutation while iterating, as that is probably a common use case for the iteration function in the first place.

Reply:
If it is possible to allow mutation during capcon iteration it is probably also possible to allow mutation during regular storage iteration as well then, considering they both iterate over a storage domain (if i understand correctly). We should look into relaxing that restriction if we can get this to work.

Reply:
I asked for some feedback on the desired behaviour on the FLIP: https://github.com/onflow/flow/pull/798#issuecomment-1542632876

Reply:
> If it is possible to allow mutation during capcon iteration it is probably also possible to allow mutation during regular storage iteration as well then, considering they both iterate over a storage domain (if i understand correctly).

It is possible, by taking a snapshot of all keys. This is potentially very (or even prohibitively) expensive when there are many entries (here, many controllers). As the point of the iteration functions is to avoid such expensive allocations, still snapshotting "under-the-hood" is probably not an option, but I've asked for feedback on it.